[WIP] Migration guide for providers to implement V2 primitives (#2496)
Qiskit/documentation: docs/migration-guides/qiskit-backend-primitives.mdx

---
title: Migrate to primitives for users of third-party providers with backend.run
description: Migrate to primitives for users of third-party providers with `backend.run`

---

# Migrate to primitives for users of third-party providers with `backend.run`

> The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third-party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run` remains in Qiskit for backward compatibility.

Reviewer: ```suggestion
The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third-party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (`Sampler` and `Estimator`) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run` remains in Qiskit for backward compatibility.
```

`PrimitiveResult` interface. See the [Primitive inputs and outputs](/guides/primitive-input-output) for more details.

The two main differences with the implementations of [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2) are:

1.  The `backend` input passed during the "backend primitive" instantiation. This input argument allows access 
to the internal provider through `backend.run()` and should not be confused
with the `mode` input argument of [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2).

2. What `options` can be set. Options should be provided as a `dict` instead of an `Options` instance.
`BackendSamplerV2` includes `default_shots`, `seed_simulator` and `run_options`;
`BackendEstimatorV2` includes `default_precision`, `abelian_grouping`, `seed_simulator` and `run_options`.
The `run_options` field defines the options to pass through to the `.run()` method of the wrapped backend instance.

Note that because these wrappers are independent of `qiskit_ibm_runtime`, they do not offer access to execution

> modes (`Session`/`batch`), local testing modes, or any other `qiskit_ibm_runtime`-specific feature. These are up to the

Reviewer: ```suggestion
modes (`Session`/`Batch`), local testing modes, or any other `qiskit_ibm_runtime`-specific feature. These are up to the
```

---
title: Migrate to primitives for third-party provider users
description: Migrate to primitives for users of third-party providers with `backend.run`

---

# Migrate to primitives for users of third-party providers with `backend.run`

> The `backend.run()` interface, previously the primary access point for IBM&reg; QPUs, has been superseded 
> by the Qiskit [primitives interface](/guides/primitives) in `qiskit_ibm_runtime`. 
> While this has become the standard for IBM providers, not all third-party providers have transitioned 
> to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.

Reviewer: Let's focus this on it being a change in standard interface defined in Qiskit, rather than a change in IBM Quantum.

```suggestion
The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.
```

This guide is for you if you have written or want to write code using `Sampler` and `Estimator` 
primitives but need to interact with a provider that still relies on `backend.run()`. If you are a 
provider developer looking to implement and expose your own `Sampler` and 
`Estimator` primitives, you can find guidance in the [Migrate provider interfaces from backend.run to primitives](/migration-guides/external-providers-primitives-v2) guide.  

For providers that don't expose native primitives, the Qiskit SDK offers wrappers for `backend.run` that you can 
use out-of the box: the [`BackendSamplerV2`.](/api/qiskit/qiskit.primitives.BackendSamplerV2) and 
[`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) classes.

If you are already familiar with the [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2) interfaces, the transition to 
[`BackendSamplerV2`.](/api/qiskit/qiskit.primitives.BackendSamplerV2) and 
[`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) is seamless, as
these classes follow the same usage patterns. The inputs to the primitives should follow 

> the Primitive Unified Bloc (PUB) syntax, and the output formats should follow the
> V2 primitives standard. See the [Overview of PUBs section](/guides/primitive-input-output#pubs) in 
> the Primitive inputs and outputs guide for details.

Reviewer: ```suggestion
the Primitive Unified Bloc (PUB) syntax, and the output formats should follow the
`PrimitiveResult` interface. See the [Primitive inputs and outputs](/guides/primitive-input-output) for more details.
```

The two main differences with the implementations of [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2) are:

> 1.  The `backend` input passed during the "backend primitive" instantiation. This input argument allows to 
> internally wrap the access to the internal provider through `backend.run()` and it should not be confused

Reviewer: ```suggestion
1.  The `backend` input passed during the "backend primitive" instantiation. This input argument allows access to the internal provider through `backend.run()` and should not be confused
```

with the `mode` input argument of [`SamplerV2`](/api/qiskit-ibm-runtime/sampler-v2) 
and [`EstimatorV2`](/api/qiskit-ibm-runtime/estimator-v2).

> 2. The use of `options`, which contains one more level of nesting compared to provider-specific primitives and

Reviewer: I don't understand where this additional level of nesting comes from, given you can also do `options={"default_shots": 1024}` using qiskit-ibm-runtime.

Author: Additional level of nesting is probably not the right term, I wanted to find a nice way to explain that the structure of the options is sometimes different. The level of nesting point referred to `run_options` which may bundle options that are "first level" in the runtime primitives. But I might be confusing things. It might be better to simply lay out how the options look like. The paragraph below does explain that custom option setting might require a subclass anyways.

is initially provided as a `dict` instead of an `Options` instance.
In the first level of nesting, `BackendSamplerV2` includes `default_shots`, `seed_simulator` and `run_options`;
`BackendEstimatorV2` includes `default_precision`, `abelian_grouping`, `seed_simulator` and `run_options`.
The `run_options` field defines the options to pass through to the `.run()` method of the wrapped backend instance.

Note that because these wrappers are independent of `qiskit_ibm_runtime`, they do not offer access to execution
modes (`Session`/`batch`), local testing modes, or any other `qiskit_ibm_runtime`-specific feature. These are up to the
third-party provider to implement and expose, ideally through a native primitive. 
If the use of the third-party backend requires a more fine-grained option handling
strategy that is not covered by the out-of-the-box wrapper, it is always possible to extend the 
[`BackendSamplerV2`.](/api/qiskit/qiskit.primitives.BackendSamplerV2) and 
[`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) interfaces through subclassing to fit the specific 
interface needs.

> The following is an example usage of the [`BackendEstimatorV2`](/api/qiskit/qiskit.primitives.BackendEstimatorV2) 

Reviewer: The code uses a sampler, not estimator

```suggestion
The following is an example usage of the [`BackendSamplerV2`](/api/qiskit/qiskit.primitives.BackendSamplerV2) 
```

class with a fictional third-party provider backend that only implements `backend.run()`:

``` python
from qiskit import QuantumCircuit, generate_preset_pass_manager
from qiskit.primitives import BackendSamplerV2
from qiskit_fictional_provider import BackendRunProvider

# get backend and pass onto BackendSamplerV2
backend = BackendRunProvider("").get_backend("fictional_backend")
sampler = BackendSamplerV2(backend, options={"default_shots": 1000})

# from this point onwards, use sampler as any SamplerV2
qc = QuantumCircuit(2)
qc.cx(0, 1)
qc.measure_all()

# assuming that third-party backend is compatible with our transpilation pipeline
# (whether this step is required or not will depend on the provider)
pm = generate_preset_pass_manager(backend=backend)
tqc = pm.run(qc)

# run using pub syntax

> result = sampler.run(([qc])).result()

Reviewer: The additional `()` doesn't do anything. If you want to make it explicitly a tuple, it needs to be `run(([qc],))`

```suggestion
result = sampler.run([qc]).result()
```

# Migrate to primitives for users of third-party providers with `backend.run`

> The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.

Reviewer 2: ```suggestion
The `backend.run` interface was the primary access point to quantum hardware and simulators offered by third-party providers that integrate with Qiskit. The evolution of [user access patterns for QPUs](/migration-guides/qiskit-runtime#why-migrate-to-qiskit-runtime-primitives) has led to the Qiskit primitives (Sampler and Estimator) superseding this interface. However, not all third-party providers have transitioned to the primitives framework, and `backend.run()` remains in Qiskit for backward compatibility.
```
